### ISSUE TITLE
Missing toggle function: DoMatchParen NoMatchParen

### ISSUE THREAD

**Issue by kiryph** [andymass/vim-matchup#5]
The original plugin [matchparen.vim](https://github.com/vim/vim/blob/master/runtime/plugin/matchparen.vim) provides two functions
```viml
command! DoMatchParen call s:DoMatchParen()
command! NoMatchParen call s:NoMatchParen()
```

I am using them in the following way inspired by `vim-unimpaired` by defining in my vimrc:
```viml
augroup vimenter
  autocmd!
  autocmd VimEnter * NoMatchParen
augroup END
nnoremap [op :DoMatchParen<CR>
nnoremap ]op :NoMatchParen<CR>
nnoremap cop :<C-R>=exists("g:loaded_matchparen") ? 'NoMatchParen' : 'DoMatchParen'<CR><CR>
```

I enable highlighting of matching objects (parenthesis or b:match_words) only on demand with `cop`. I usually consider them too slow, however, sometimes it is good to see matching objects.

BTW, are you aware of the pull request https://github.com/vim/vim/pull/1338?

**Comment by andymass** [comment]
Are you requesting that commands `DoMatchParen` and `NoMatchParen` to affect match-it, or are you requesting similar commands, for example, `DoMatchupMatchParen` or  `NoMatchupMatchParen`?

Do you know you can freely toggle highlighting globally with `g:matchup_matchparen_enabled = 0` and per buffer as `let b:matchup_matchparen_enabled=0` and `let b:matchup_matchparen_fallback=0`?  But I am willing to add shortcut commands; should be an easy change.

> BTW, are you aware of the pull request..

Funny!  I was not aware but I actually already implemented something similar.  I have been waiting on releasing it for a few reasons.  1) I find the slight delay annoying personally, and match-it hasn't been too slow for me to warrant it. 2) It adds another option which I don't want to make the default but then people won't be aware of it. 3) `has('timers')` is pretty new.  Tracking progress here: #6.

**Comment by kiryph** [comment]
> Do you know you can freely toggle highlighting globally with g:matchup_matchparen_enabled = 0 and per buffer as let b:matchup_matchparen_enabled=0 and let b:matchup_matchparen_fallback=0?

I am not sure that changing the value of the variable enables dynamically matchup_matchparen.
I mean I have added to my vimrc `let g:matchup_matchparen_enabled = 0` so matchup is disabled on startup. However, I want to enable it with `cop` and switch it off again with `cop`.

Following toggles the value of the variable but does not enable or disable matchup_matchparen:

vimrc
```
...
let g:matchup_matchparen_enabled = 0
nnoremap cop :<C-R>=exists("g:matchup_matchparen_enabled") ? (g:matchup_matchparen_enabled == 1? 'let g:matchup_matchparen_enabled = 0': 'let g:matchup_matchparen_enabled = 1'): 'let g:matchup_matchparen_enabled = 0'<CR><CR>
...
```

**Comment by andymass** [comment]
Ah, I see the issue with `g:matchup_matchparen_enabled` now.

I've added DoMatchParen and NoMatchParen which should do what you want.

**Comment by kiryph** [comment]
Thanks for the update. My setup if I have matchup installed is now:

```viml
" let g:matchup_matchparen_enabled = 1 "(default) enable on startup and disable it immediately again with autocmd VimeEnter * NoMatchParen
let g:matchup_matchparen_fallback = 0 " never use matchparen.vim
augroup vimenter
  autocmd!
  autocmd VimEnter * NoMatchParen
augroup END
nnoremap [op :DoMatchParen<CR>
nnoremap ]op :NoMatchParen<CR>
" commented matchparen.vim version:
" nnoremap cop :<C-R>=exists("g:loaded_matchparen") ? 'NoMatchParen' : 'DoMatchParen'<CR><CR>
nnoremap cop :<C-R>=exists("g:matchup_matchparen_enabled") ? ( (g:matchup_matchparen_enabled == 1) ? 'NoMatchParen': 'DoMatchParen'): 'DoMatchParen'<CR><CR>
```

It does what I wanted.

I just want to say that moving by keeping hjkl pressed with matchup enabled is really slow, not better than the original matchparen.vim

**Comment by andymass** [comment]
> with matchup enabled is really slow, not better than the original matchparen.vim

match-up is necessarily going to be slower at highlighting than matchparen since it's looking for many more things.  However, I am happy to try to improve the performance as much as possible.  What language(s) do you see the slowdown with?

Would you be willing to try out the [feature-deferred](https://github.com/andymass/vim-matchup/tree/feature-deferred) branch, which enables lazy/deferred/delayed highlighting?  If so, you can do `let g:matchup_matchparen_deferred = 1` with that branch.  Also note that the times in milliseconds are configurable but you must reload for the new values to take effect.
